Support forward scrolling.

showell opened the pull request:
The first several commits are pure refactorings and should be ready to merge here.

timabbott commented on static/js/message_fetch.js:
```diff
@@ -128,14 +128,22 @@ exports.load_messages = function (opts) {
+            // small narrows we can short circuit fetching early.
+            if (messages.length < num_before + num_before - 2) {
+                fetch_status.finish_fetching();
+            }
```
I think this logic isn't correct; I think you want `min(num_before, num_after) - 1` for this algorithm.

Regardless, we can do better than this; I believe that the `load_messages` API endpoint returns an `anchor` if we didn't specify one (i..e if we used `use_first_unread_anchor`), so we always have the anchor in the return code path.  We should be able to thus compare the number of messages with an ID above or below the narrow that were returned when narrowing.  

(Or we could just do that message ID counting logistics in the backend and send the result, and make life a little easier for clients, which might be an even better idea).

timabbott commented:
Merged the first 4 commits.

showell commented:
This is still mostly refactoring, although the last change should improve the home view a bit for the case that you scroll all the way back, then un-narrow, and narrow.  (It will still realize you've gotten the oldest message for the home msg_list.)

timabbott commented:
A few thoughts:
* I think we could make the `start_older_batch` type methods in `FetchStatus` throw an exception if `found_oldest` is true, just as a programming safety thing.
* Otherwise, this looks great.  I'm going to deploy on chat.zulip.org for testing and then merge.

timabbott commented:
I merged this batch of commits.

showell commented:
The first couple commits here are fine, but I'm still figuring out the home-vs.-all stuff when it comes to not appending messages to the message lists when we're still fetching.

timabbott commented:
I merged the first 3 commits.  The next few look reasonable, though I agree we need to figure out the `msg_list` append-at-bottom stuff.

showell commented:
The first ~4 here are mostly refactors or minor behavior changes.

timabbott commented:
I merged the first 4 commits.  Left the 5th since there was still some investigation being done there.

timabbott commented:
@showell is this ready for review?  I'm curious about the `blueslip.info` part in particular.

timabbott commented:
I merged the next two commits (unit tests).

showell commented:
@timabbott, This is ready for another test deploy, and then I think it's good to merge to master.

For the chrome bug fix, I assume you still have that on your branch somewhere?

Since we last test deployed this PR, the following changes have been made:
* I added node tests, which are obviously no risk.  (You merged 2 of them, but the 3rd commit depended on the "main" fix here.)
* I added logic to keep the fetch statuses in sync for message_list.all and home_msg_list.
* When a message list is behind, I now actually ignore the message as well as blueslipping it.
* Likewise for local echo, if the current list is behind, I do a blueslip.info and skip local echo  (which is always just a luxury, anyway).

The only thing that feels vaguely risky here is ignoring new messages as they come in, when the lists are behind, so I focused most of my manual testing on that.  With two browsers, you can clearly see the blueslip when the narrow is behind, and then you can hit end to catch up and the recent message will be there.  I didn't try to simulate the race condition, but I think worst case there is one message gets missed.

I also, with just one user, sent messages to the stream in far-behind narrows, and you see two blueslips there (one for skipping echo, one for ignore the "new" message from the server), and then you'd see the message when you caught up.

timabbott commented on static/js/message_list.js:
```diff
@@ -42,6 +42,11 @@ exports.MessageList.prototype = {
         var interior_messages = [];
 
+        if (opts.messages_are_new && !self.fetch_status.can_append()) {
+            blueslip.info('ignoring new message for list that is behind on fetches');
```
Should this be `blueslip.log` or lower?  I.e. do we want this in the JS console?

timabbott commented on static/js/echo.js:
```diff
@@ -146,6 +146,11 @@ exports.try_deliver_locally = function try_deliver_locally(message_request) {
     }
 
+    if (!current_msg_list.fetch_status.can_append()) {
+        blueslip.info('suppressing echo for list that is behind on fetches');
```
This one should probably be `blueslip.log`.

showell commented:
I'm fine to change both of those to log.  Is there anything else to address?

timabbott commented:
I think that's it.  I merged ac7031de3a93d276fbcdcad3c9481de29839543a, which was the only known issue, though we should of course debug whatever Rishi reported today.

timabbott commented:
Merged, after reordering and tagging the key commit to close #3465.  Thanks @showell!
